### PR TITLE
dns: configuring a basic key value store to persist to disk

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low (config guarded)
Testing: new unit, integration tests
Docs Changes: n/a (hidden)
Release Notes: n/a
Part of https://github.com/envoyproxy/envoy-mobile/issues/1587

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
Would this overall feature make sense as part of the DNS extensibility being worked on in https://github.com/envoyproxy/envoy/pull/17479?

I.e. we add a caching DNS resolver extension (that maybe wraps another DNS resolver like c-ares or platform specific)?

**Comment by mattklein123** [comment]
> I.e. we add a caching DNS resolver extension (that maybe wraps another DNS resolver like c-ares or platform specific)?

I think this would potentially be a nice solution to think about it. FWIW we had discussed something similar today for wrapping DNS in order to get DNS specific analytic events. cc @snowp

**Comment by alyssawilk** [comment]
I'm not sure what the ask is for this PR.  I don't want this tied to DNS but I think it'd be reasonable to make sure the DNS refactor allows for all the platform resolvers to use the cache.

**Comment by htuch** [comment]
i.e. that we build DNS caching inside the DNS resolver framework. I'm not sure why we want to reuse the existing data plane caching framework for this.

File: api/envoy/extensions/cache/key_value_cache/v3/config.proto [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// [#protodoc-title: Key Value Cache storage plugin]
+
+// This shared configuration for Envoy key value stores.
+message KeyValueCacheConfig {
+  // Extensible config for key values caches.
+  // The expectation is that these are flushed to disk but for mobile, that may
+  // be via indirect APIs so may need more than just a simple file name.
+  string name = 1 [(validate.rules).string = {min_len: 1}];
+
+  google.protobuf.Any typed_config = 2;
```
Comment (htuch) [inline api/envoy/extensions/cache/key_value_cache/v3/config.proto:25]: This shold be `TypedExtensionConfig`

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 2]
```diff
@@ -413,6 +419,48 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
...
+  key_value_store_ = factory.createStore(config.persistent_cache_config(), validation_visitor_,
+                                         main_thread_dispatcher_, file_system_);
+  KeyValueStore::ConstIterateCb load = [this](const std::string& key, const std::string& value) {
+    auto address = Network::Utility::parseInternetAddressAndPortNoThrow(value);
+    if (address == nullptr) {
+      return KeyValueStore::Iterate::Break;
+    }
+    stats_.cache_load_.inc();
+    std::list<Network::DnsResponse> response;
+    response.emplace_back(Network::DnsResponse(address, std::chrono::seconds(0) /* ttl */));
```
Comment (RyanTheOptimist) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:454]: By passing in `seconds(0)` does this mean the cache entries are already expired?

**Review by RyanTheOptimist — COMMENTED** [review]
A few small questions to get started...

File: api/envoy/extensions/cache/key_value_cache/v3/config.proto (hunk 1 above)
Comment (alyssawilk) [reply api/envoy/extensions/cache/key_value_cache/v3/config.proto:25]: done

File: source/extensions/cache/key_value_store/file_based/config.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+KeyValueStorePtr FileBasedKeyValueStoreFactory::createStore(
+    const Protobuf::Message& config, ProtobufMessage::ValidationVisitor& validation_visitor,
+    Event::Dispatcher& dispatcher, Filesystem::Instance& file_system) {
+  const auto& typed_config = MessageUtil::downcastAndValidate<
+      const envoy::extensions::cache::key_value_cache::v3::KeyValueCacheConfig&>(
+      config, validation_visitor);
+  const auto file_config = MessageUtil::anyConvertAndValidate<
+      envoy::extensions::cache::key_value_cache::v3::FileBasedKeyValueCacheConfig>(
+      typed_config.typed_config(), validation_visitor);
+  auto ms = std::chrono::seconds(DurationUtil::durationToSeconds(typed_config.flush_interval()));
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/extensions/cache/key_value_store/file_based/config.cc:53]: yeah that was my bad - should be seconds thanks.

File: source/extensions/cache/key_value_store/file_based/config.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+namespace Extensions {
+namespace Cache {
+namespace KeyValueCache {
+
+// A filesystem based key value store, which loads from and flushes to the file
+// provided.
+//
+// All keys and values are flushed to a single file as
+// [length]\n[key][length]\n[value]
+class FileBasedKeyValueStore : public KeyValueStoreBase {
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/extensions/cache/key_value_store/file_based/config.h:17]: yeah, per offline conversation with harvey I moved all the things to a location he prefered, so now it's KeyValue and then KeyValueStore which is at least a substring?

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 5]
```diff
@@ -413,6 +419,48 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
   parent_.stats_.num_hosts_.dec();
 }
 
+void DnsCacheImpl::addCacheEntry(const std::string& host,
+                                 const Network::Address::InstanceConstSharedPtr& address) {
+  if (!key_value_store_) {
+    return;
+  }
+  const std::string value = absl::StrCat(address->asString());
+  key_value_store_->addOrUpdate(host, value);
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:428]: oh sorry, I could pass in the string view, but I plan on adding TTL.  So this will be used for temporary storage in the immediate follow up and I'd prefer to leave as is?

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 2 above)
Comment (alyssawilk) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:454]: AFIK ttl in this struct is not used :-(
DNS TTL is filled in, and ignored in favor of the "how long do we persist cache entries" hard coded config.  IMO DNS should persist a max of server-returned TTL and configured "longest time I'm willing to wait" but given it's ignored and a I have a TODO to handle lifetimes I left this for later.

**Review by alyssawilk — COMMENTED** [review]
FYI there's an outstanding issue with both clang tidy and docs build, but API should be ready for review.  PTAL.

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto [hunk 6]
```diff
@@ -138,4 +139,8 @@ message DnsCacheConfig {
   // Setting this timeout will ensure that queries succeed or fail within the specified time frame
   // and are then retried using the standard refresh rates. Defaults to 5s if not set.
   google.protobuf.Duration dns_query_timeout = 11 [(validate.rules).duration = {gt {}}];
+
+  // [#not-implemented-hide:]
+  // Configuration to flush the DNS cache to long term storage.
+  key_value.v3.KeyValueStoreConfig persistent_cache_config = 13;
```
Comment (htuch) [inline api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:145]: Nit on naming: given this is referencing an extension point, do we know that there will always be persistence guaranteed?

File: api/envoy/extensions/common/key_value/v3/config.proto [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+// [#alpha:]
+// This shared configuration for Envoy key value stores.
+message KeyValueStoreConfig {
+  // [#extension-category: envoy.common.key_value]
+  config.core.v3.TypedExtensionConfig config = 1 [(validate.rules).message = {required: true}];
+
+  // The interval at which the key value store should be flushed to long term
+  // storage.
+  google.protobuf.Duration flush_interval = 2;
```
Comment (htuch) [inline api/envoy/extensions/common/key_value/v3/config.proto:28]: I guess this gets to my earlier question, do we know that all implementations will flush? Or will flushing be a property of a specific key-value story, e.g. the `FileBasedKeyValueStoreConfig`?

File: api/envoy/extensions/common/key_value/v3/config.proto [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+// [#alpha:]
+// [#extension: envoy.common.key_value.file_based]
+// This is configuration to flush a key value store out to disk.
+message FileBasedKeyValueStoreConfig {
```
Comment (htuch) [inline api/envoy/extensions/common/key_value/v3/config.proto:34]: This should live in `api/envoy/key_value/file_based/v3` or something I think. It seems a lot of boiler plate, but we have some negative counter-examples of situations where extension points and their extensions are grouped as monolith files, see stats sinks. Best practice is to put extension instances in their own hierarchy.

**Review by htuch — COMMENTED** [review]
API looks good modulo comments.

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 2 above)
Comment (RyanTheOptimist) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:459]: *blink* Well that's really something! :) For Envoy as a proxy, I can imagine this is no problem. I won't if we'll end up with any wrinkles in the context of Envoy as a client if we're caching DNS names longer than the advertised TTL?

But back to this PR, perhaps we'll end up wanting to persist the cache expiration time to disk so that when we reload them we can evict any entries that have already expired? (Maybe something to do when you resolve the todo below?)

**Review by RyanTheOptimist — APPROVED** [review]
Looks good!

**Review by alyssawilk — COMMENTED** [review]
looking at docs pr - must have failed a string somewhere

File: api/envoy/extensions/common/key_value/v3/config.proto (hunk 8 above)
Comment (htuch) [reply api/envoy/extensions/common/key_value/v3/config.proto:30]: Great. Now that you've moved to a distinct file, I assume this can be deleted?

File: api/envoy/extensions/common/key_value/file_based/v3/config.proto [hunk 9]
```diff
@@ -0,0 +1,27 @@
+syntax = "proto3";
+
+package envoy.extensions.common.key_value.file_based.v3;
```
Comment (htuch) [inline api/envoy/extensions/common/key_value/file_based/v3/config.proto:3]: Nit: this should be in `api/envoy/extensions/key_value/filed_based/v3/config.proto` and same package namespace.

The rationale is that the `KeyValueConfig` is a common extension point declaration reused across protos. The `FileBasedKkeyValueStoreConfig` is a `key_value` extension, and all extension classes have a root like `api/envoy/extensions/<extension class>`.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by alyssawilk** [comment]
argh sorry, force pushed format fixes.  Bad @alyssawilk no biscuit

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 10]
```diff
@@ -413,6 +423,49 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
...
+  if (!config.has_key_value_config()) {
+    return;
+  }
+  auto& factory =
+      Config::Utility::getAndCheckFactory<KeyValueStoreFactory>(config.key_value_config().config());
+  key_value_store_ = factory.createStore(config.key_value_config(), validation_visitor_,
+                                         main_thread_dispatcher_, file_system_);
+  KeyValueStore::ConstIterateCb load = [this](const std::string& key, const std::string& value) {
+    auto address = Network::Utility::parseInternetAddressAndPortNoThrow(value);
+    if (address == nullptr) {
```
Comment (snowp) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:454]: Should we log/inc a stat in this case? This means that the cached data is bad right?

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 11]
```diff
@@ -413,6 +423,49 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
...
+      return KeyValueStore::Iterate::Break;
+    }
+    stats_.cache_load_.inc();
+    std::list<Network::DnsResponse> response;
+    // TODO(alyssawilk) change finishResolve to actually use the TTL rather than
+    // putting 0 here, return the remaining TTL or indicate the result is stale.
+    response.emplace_back(Network::DnsResponse(address, std::chrono::seconds(0) /* ttl */));
+    startCacheLoad(key, address->ip()->port());
```
Comment (snowp) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:462]: I'm not that familiar with this code, but it looks like this will kick off a DNS query? And then we immediately resolve it? 

If so I wonder if this will cause unforeseen consequences, like having entries being resolved twice (once from cache and once from DNS) and confusing stats (e.g. stat increment for DNS resolution and cache load for same entry)

Thoughts?

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc [hunk 12]
```diff
@@ -1006,6 +1024,98 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
...
+        envoy::extensions::key_value::file_based::v3::FileBasedKeyValueStoreConfig>();
+  }));
+  MockKeyValueStore* store{};
+  EXPECT_CALL(factory, createStore(_, _, _, _)).WillOnce(Invoke([&store]() {
+    auto ret = std::make_unique<NiceMock<MockKeyValueStore>>();
+    store = ret.get();
+    // Make sure there's an attempt to load from the key value store.
+    EXPECT_CALL(*store, iterate);
+    return ret;
+  }));
```
Comment (snowp) [inline test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:1041]: Should this be implemented on the MockKeyValueStoreFactory itself? Having them in the test case here is a bit distracting from what the test is actually testing

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc [hunk 13]
```diff
@@ -1006,6 +1024,98 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
...
+
+  Registry::InjectFactory<KeyValueStoreFactory> injector(factory);
+  config_.mutable_key_value_config()->mutable_config()->set_name("mock_key_value_store_factory");
+
+  initialize();
+  InSequence s;
+  ASSERT(store != nullptr);
+
+  // Make sure the store gets the first insert.
```
Comment (snowp) [inline test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:1050]: It's a bit hard to follow how the below code block verifies anything with the store, maybe make that clearer with comments? (I think it's L1067 but there is a lot of boiler plate here)

**Review by snowp — CHANGES_REQUESTED** [review]
Overall this looks pretty good, just a few nits and a question on the interaction between cache load and DNS resolution

File: envoy/common/key_value_store.h [hunk 14]
```diff
@@ -37,6 +41,45 @@ class KeyValueStore {
...
+   * @param cb supplies the key value store configuration
+   * @param validation_visitor the configuration validator
+   * @dispatcher the dispatcher for the thread, for flush alarms.
+   * @file_system the file system.
+   * @return a new key value store.
+   */
+  virtual KeyValueStorePtr createStore(const Protobuf::Message& config,
+                                       ProtobufMessage::ValidationVisitor& validation_visitor,
+                                       Event::Dispatcher& dispatcher,
+                                       Filesystem::Instance& file_system) PURE;
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply envoy/common/key_value_store.h:79]: Unfortunately we can't - the DNS store is created both from ServerFactoryContext and from the cluster's TransportSocketFactoryContextImpl which have wide overlap but no common base class.
I have that filed in away in my clean up list but I think resolving it is out of scope for this PR

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 11 above)
Comment (alyssawilk) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:462]: it works today (see integration test) but I agree it's not great, hence the TODO :-)

What I want to have happen is be able to differentiate stale cached results from fresh.  I do want to immediately kick off a resolve, but also have stale data in there, config for how long we wait for fresh data before we use stale data.  I plan to do that in the follow up

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc (hunk 12 above)
Comment (alyssawilk) [reply test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:1041]: maybe?  Honestly I regularly get bit by expectations in mock timer where I don't understand what's going on, so I'm hesitant to add complexity like that to the mock base classes

**Comment by alyssawilk** [comment]
@davinci26 @wrowe any ideas on the windows failure?  I've commented out with TODO to unblock this but I'm wondering if you have ideas what went wrong. For example are there known gotchas with file paths on windows CI?

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 11 above)
Comment (snowp) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:463]: For my understanding, would a DNS resolution then overwrite this value once it arrives?

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc (hunk 12 above)
Comment (snowp) [reply test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:1041]: Sure not a big deal, just what I usually see done with mocks. Agree that MockTimer is especially confusing :)

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 11 above)
Comment (alyssawilk) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:463]: yeah, fresh resolve always overwrites older data (if it's different)
in this case it'd be the same so not trigger any changes.

**Review by snowp — APPROVED** [review]
LG! We definitely need to figure out the Windows thing, but I'm okay with doing that in a follow up

**Comment by davinci26** [comment]
@alyssawilk I am not sure, I will have to debug locally but it is super weird because the same test passes on windows with clang.

**Comment by alyssawilk** [comment]
coverage flake unrelated: from a combination of quic refactor and setec patches
I've got https://github.com/envoyproxy/envoy/pull/17846 to hep so I'm going to merge over it.
